elementpath/xpath_nodes.py

```diff
@@ -136,480 +161,755 @@ class NamespaceNode(XPathNode):
-                return uri
-    return None
+            if not callable(elem.tag):
+                child = build_lxml_element_node()
+            elif elem.tag.__name__ == 'Comment':  # type: ignore[attr-defined]
+                child = CommentNode(elem, parent, position)
+                position += 1
+            else:
+                child = ProcessingInstructionNode(elem, parent, position)
 
+            parent.children.append(child)
+            if elem.tail is not None:
+                parent.children.append(TextNode(elem.tail, parent, position))
+                position += 1
 
-def node_children(obj: Any) -> Optional[Iterator[ElementNode]]:
-    if is_element_node(obj):
-        return (child for child in obj)
-    elif is_document_node(obj):
-        return (child for child in [obj.getroot()])
-    else:
-        return None
+            if len(elem):
+                ancestors.append(parent)
+                parent = child
+                iterators.append(children)
+                children = iter(elem)
 
 
-def node_nilled(obj: Any) -> Optional[bool]:
-    if is_element_node(obj):
-        if isinstance(obj, TypedElement):
-            return obj.elem.get(XSI_NIL) in ('true', '1')
-        return obj.get(XSI_NIL) in ('true', '1')
-    return None
-
-
-def node_kind(obj: Any) -> Optional[str]:
-    if isinstance(obj, XPathNode):
-        return obj.kind
-    elif is_element_node(obj):
-        return 'element'
-    elif is_document_node(obj):
-        return 'document-node'
-    elif is_comment_node(obj):
-        return 'comment'
-    elif is_processing_instruction_node(obj):
-        return 'processing-instruction'
-    else:
-        return None
+def build_schema_node_tree(root: Union[XsdElementProtocol, XMLSchemaProtocol]) -> ElementNode:
+    root_node: Union[DocumentNode, ElementNode]
+    parent: Any
+    elem: Any
+    child: ChildNodeType
+    children: Iterator[Any]
 
+    position = 1
 
-def node_name(obj: Any) -> Optional[str]:
-    if isinstance(obj, XPathNode):
-        return cast(Optional[str], obj.name)
-    elif not hasattr(obj, 'tag') or not hasattr(obj, 'text'):
-        return None
-    elif not callable(obj.tag):
-        return cast(str, obj.tag)
-    elif obj.tag.__name__ != 'ProcessingInstruction':
-        return None
-    else:
-        # Return pi target. ElementTree doesn't have a specific attribute
-        # for target but put it before the text, separated by a space.
+    def build_schema_element_node(xsd_type: Optional[XsdTypeProtocol] = None) -> ElementNode:
+        nonlocal position
+
+        node = ElementNode(elem, parent, position, root.namespaces, xsd_type)
+        position += 1
+
+        if elem.attributes:
+            node.attributes = []
+            for name, attr in elem.attributes.items():
+                node.attributes.append(
+                    AttributeNode(name, attr, node, position, attr.type)
+                )
+                position += 1
```

Comment:
I've seen this code several times. Seems worth a helper function. (Note that increasing `position` by 1 inside of the loop is the same as adding the number of attributes to it after the loop. `enumerate()` can be used for the loop instead, even in a list comprehension.)

Reply:
I was thinking of something like this:
```python
def collect_attributes(elem, initial_position=0):
    return [
        AttributeNode(name, attr, node, pos, attr.type)
        for pos, (name, attr) in enumerate(elem.attributes.items(), initial_position)
    ]

node.attributes = collect_attributes(elem, position)
position += len(node.attributes)
```